code/modules/reagents/chemistry/machinery/chem_recipe_debug.dm

```diff
@@ -105,25 +153,207 @@
+			reagents.trans_to(B)
+			B.name = "[cached_reactions[index]]"
 		reagents.clear_reagents()
+		reagents.chem_temp = 300
 		if(failed > 1)
 			index++
 			failed = 0
-		if(failed == 0)
-			index++
+	else
+		say("No reagents left in beaker!")
+
+/obj/machinery/chem_recipe_debug/proc/setup_reaction()
+	react_time = 0
+	if(!length(cached_reactions))
+		return FALSE
 	var/datum/chemical_reaction/C = cached_reactions[index]
 	if(!C)
 		say("Unable to find reaction on index: [index]")
+	say("Using forced temperatures.")
+	if(C.reaction_flags & REACTION_INSTANT)
+		say("This reaction is instant")
 	for(var/R in C.required_reagents)
-		reagents.add_reagent(R, C.required_reagents[R]*20)
+		reagents.add_reagent(R, C.required_reagents[R]*vol_multi)
 	for(var/cat in C.required_catalysts)
 		reagents.add_reagent(cat, C.required_catalysts[cat])
-	if(failed == 0)
+	if(force_temp && !min_temp)
+		say("Using forced temperatures.")
+		reagents.chem_temp = f_temp ? f_temp : C.optimal_temp
+	if(force_ph)
+		say("Using forced pH.")
+		reagents.ph = f_ph ? f_ph : (C.optimal_ph_max + C.optimal_ph_min)/2
+	if(failed == 0 && !force_temp)
 		reagents.chem_temp = C.optimal_temp
-	if(failed == 1)
+	if(failed == 1 && !force_temp)
 		reagents.chem_temp = C.required_temp+25
 		failed++
+	if(min_temp)
+		say("Overriding temperature to required temp.")
+		reagents.chem_temp = C.is_cold_recipe ? C.required_temp - 1 : C.required_temp + 1
 	say("Reacting <span class='nicegreen'>[cached_reactions[index]]</span> starting pH: [reagents.ph] index [index] of [cached_reactions.len]")
-	if(C.reaction_flags & REACTION_INSTANT)
-		say("This reaction is instant")
 
+/obj/machinery/chem_recipe_debug/ui_data(mob/user)
+	var/data = list()
+	data["targetTemp"] = f_temp
+	data["targatpH"] = f_ph
+	data["isActive"] = reagents.is_reacting
+	data["forcepH"] = force_ph
+	data["forceTemp"] = force_temp
+	data["targetVol"] = vol_multi
+	data["processAll"] = process_all
+	data["currentTemp"] = reagents.chem_temp
+	data["currentpH"] = round(reagents.ph, 0.01)
+	data["processing"] = processing
+	data["index"] = index
+	data["endIndex"] = cached_reactions.len
+	data["beakerSpawn"] = beaker_spawn
+	data["minTemp"] = min_temp
+
+	var/list/beaker_contents = list()
+	for(var/r in reagents.reagent_list)
+		var/datum/reagent/reagent = r
+		beaker_contents.len++
+		beaker_contents[length(beaker_contents)] = list("name" = reagent.name, "volume" = round(reagent.volume, 0.01))
+	data["chamberContents"] = beaker_contents
+
+	var/list/queued_reactions = list()
+	for(var/datum/chemical_reaction/react as anything in reaction_names)
+		var/datum/reagent/R = find_reagent_object_from_type(react.results[1])
+		queued_reactions.len++
+		queued_reactions[length(queued_reactions)] = list("name" = R.name)
+	data["queuedReactions"] = queued_reactions
+
+	var/list/active_reactions = list()
+	var/flashing = 14 //for use with alertAfter - since there is no alertBefore, I set the after to 0 if true, or to the max value if false
```

Comment:
What's this magic number? Can you move it out to a define?

Reply:
Basically, since dials will only flash for values over a certain number, instead of below, I use this to set the min value of the flashing argument in tgui to 0 if I want it to flash, or 14 if I don't want it to flash. I used this elsewhere too, but I can move the define so it applies to both.